File: backend/app/schemas/cultural_place.py
"""Cultural places API schemas (WEB-MAP-03, WEB-SEARCH-02B.1)."""

from __future__ import annotations

import uuid

from pydantic import BaseModel, Field

from app.schemas.map_event import MapBboxResponse


class CulturalPlaceNeighborhoodSummary(BaseModel):
    slug: str
    display_name: str


class CulturalGalleryImage(BaseModel):
    url: str
    alt: str | None = None
    credit: str | None = None
    source: str | None = None


class CulturalPlaceMediaFields(BaseModel):
    """Shared media payload for list, detail, and map responses."""

    image_url: str | None
    hero_image_url: str | None
    thumbnail_image_url: str | None
    gallery_images: list[CulturalGalleryImage] = Field(default_factory=list)
    editorial_excerpt: str | None
    photo_credit: str | None
    image_source: str | None


class CulturalPlaceListItem(CulturalPlaceMediaFields):
    id: uuid.UUID
    slug: str
    name: str
    short_description: str
    city: str
    address: str
    category: str
    latitude: float
    longitude: float
    image_alt: str | None
    source_name: str
    image_credit: str | None
    neighborhood: CulturalPlaceNeighborhoodSummary | None = None


class CulturalPlaceDetail(CulturalPlaceListItem):
    description: str | None
    source_url: str | None
    image_license: str | None
    is_featured: bool
    featured_priority: int
    image_blurhash: str | None = None


class CulturalPlaceListResponse(BaseModel):
    city: str
    items: list[CulturalPlaceListItem]
    count: int


class MapCulturalPlaceItem(CulturalPlaceMediaFields):
    id: uuid.UUID
    slug: str
    name: str
    category: str
    address: str
    city: str
    neighborhood: CulturalPlaceNeighborhoodSummary | None = None
    latitude: float
    longitude: float
    image_alt: str | None
    source_name: str
    image_credit: str | None


class MapCulturalPlaceListResponse(BaseModel):
    city: str
    bbox: MapBboxResponse
    count: int
    places: list[MapCulturalPlaceItem]

File: frontend/packages/types/src/cultural-place.ts
/** Cultural places API types (WEB-MAP-03, WEB-SEARCH-02B.1). */

export type CulturalPlaceNeighborhoodSummary = {
  slug: string;
  display_name: string;
};

export type CulturalGalleryImage = {
  url: string;
  alt?: string | null;
  credit?: string | null;
  source?: string | null;
};

export type CulturalPlaceMediaFields = {
  image_url: string | null;
  hero_image_url: string | null;
  thumbnail_image_url: string | null;
  gallery_images: CulturalGalleryImage[];
  editorial_excerpt: string | null;
  photo_credit: string | null;
  image_source: string | null;
};

export type CulturalPlaceListItem = CulturalPlaceMediaFields & {
  id: string;
  slug: string;
  name: string;
  short_description: string;
  city: string;
  address: string;
  category: string;
  latitude: number;
  longitude: number;
  image_alt: string | null;
  source_name: string;
  image_credit: string | null;
  neighborhood: CulturalPlaceNeighborhoodSummary | null;
  is_featured?: boolean;
  created_at?: string;
};

export type CulturalPlaceDetail = CulturalPlaceListItem & {
  description: string | null;
  source_url: string | null;
  image_license: string | null;
  is_featured: boolean;
  featured_priority: number;
  image_blurhash: string | null;
};

export type CulturalPlaceListResponse = {
  city: string;
  items: CulturalPlaceListItem[];
  count: number;
  total: number;
  offset: number;
  limit: number;
};

export type CulturalPlaceStatsResponse = {
  city: string;
  total_places: number;
  new_this_month: number;
  category_count: number;
};

export type CulturalPlaceSort = "featured" | "name" | "recent";

export type MapCulturalPlaceItem = CulturalPlaceMediaFields & {
  id: string;
  slug: string;
  name: string;
  category: string;
  address: string;
  city: string;
  neighborhood: CulturalPlaceNeighborhoodSummary | null;
  latitude: number;
  longitude: number;
  image_alt: string | null;
  source_name: string;
  image_credit: string | null;
};

export type MapCulturalPlacesListParams = {
  lat_min: number;
  lon_min: number;
  lat_max: number;
  lon_max: number;
  city: string;
  limit?: number;
};

export type MapCulturalPlaceListResponse = {
  city: string;
  bbox: {
    lat_min: number;
    lon_min: number;
    lat_max: number;
    lon_max: number;
  };
  count: number;
  places: MapCulturalPlaceItem[];
};

export type MapRouteSummary = {
  distanceMeters: number;
  durationSeconds: number;
};

export type MapRouteGeometry = {
  type: "LineString";
  coordinates: [number, number][];
};
